Mathlib/Algebra/Algebra/Bilinear.lean

```diff
@@ -0,0 +1,271 @@
+
+end NonUnital
+
+section Semiring
+
+variable (R A : Type _) [CommSemiring R] [Semiring A] [Algebra R A]
+
+/-- The multiplication in an algebra is an algebra homomorphism into the endomorphisms on
+the algebra.
+
+A weaker version of this for non-unital algebras exists as `NonUnitalAlgHom.mul`. -/
+def Algebra.lmul : A →ₐ[R] End R A :=
+  {
+    LinearMap.mul R
+      A with
+    map_one' := by
+      ext a
+      exact one_mul a
+    map_mul' := by
+      intro a b
+      ext c
+      exact mul_assoc a b c
+    map_zero' := by
+      ext a
+      exact zero_mul a
+    commutes' := by
+      intro r
+      ext a
+      exact (Algebra.smul_def r a).symm }
+#align algebra.lmul LinearMap.Algebra.lmul
+
+variable {R A}
+
+@[simp]
+theorem Algebra.coe_lmul_eq_mul : ⇑(Algebra.lmul R A) = mul R A :=
+  rfl
+#align algebra.coe_lmul_eq_mul LinearMap.Algebra.coe_lmul_eq_mul
+
+@[simp]
+theorem mulLeft_eq_zero_iff (a : A) : mulLeft R a = 0 ↔ a = 0 := by
+  constructor <;> intro h
+  -- porting note: can we avoid supplying `R` explicitly in `@mulLeft_apply` below?
+  · rw [← mul_one a, ← @mulLeft_apply R _ _ _ _ _ _ a 1, h, LinearMap.zero_apply]
+  · rw [h]
+    exact mulLeft_zero_eq_zero
+#align linear_map.mul_left_eq_zero_iff LinearMap.mulLeft_eq_zero_iff
+
+@[simp]
+theorem mulRight_eq_zero_iff (a : A) : mulRight R a = 0 ↔ a = 0 := by
+  constructor <;> intro h
+  -- porting note: can we avoid supplying `R` explicitly in `@mulRight_apply` below?
+  · rw [← one_mul a, ← @mulRight_apply R _ _ _ _ _ _ a 1, h, LinearMap.zero_apply]
+  · rw [h]
+    exact mulRight_zero_eq_zero
+#align linear_map.mul_right_eq_zero_iff LinearMap.mulRight_eq_zero_iff
+
+@[simp]
+theorem mulLeft_one : mulLeft R (1 : A) = LinearMap.id := by
+  ext
+  simp only [LinearMap.id_coe, one_mul, id.def, mulLeft_apply]
+#align linear_map.mul_left_one LinearMap.mulLeft_one
+
+@[simp]
+theorem mulRight_one : mulRight R (1 : A) = LinearMap.id := by
+  ext
+  simp only [LinearMap.id_coe, mul_one, id.def, mulRight_apply]
+#align linear_map.mul_right_one LinearMap.mulRight_one
+
+@[simp]
+theorem pow_mulLeft (a : A) (n : ℕ) : mulLeft R a ^ n = mulLeft R (a ^ n) := by
+  simpa only [mulLeft, ← Algebra.coe_lmul_eq_mul] using ((Algebra.lmul R A).map_pow a n).symm
+#align linear_map.pow_mul_left LinearMap.pow_mulLeft
+
+@[simp]
+theorem pow_mulRight (a : A) (n : ℕ) : mulRight R a ^ n = mulRight R (a ^ n) := by
+  simp only [mulRight, ← Algebra.coe_lmul_eq_mul]
+  exact
+    LinearMap.coe_injective (((mulRight R a).coe_pow n).symm ▸ mul_right_iterate a n)
+#align linear_map.pow_mul_right LinearMap.pow_mulRight
+
+end Semiring
+
+section Ring
+
+variable {R A : Type _} [CommSemiring R] [Ring A] [Algebra R A]
+
+/-- This instance should not be necessary. porting note: can we drop this? -/
+local instance : Module R A := Algebra.toModule
+
+/-- This instance should not be necessary. porting note: can we drop this? -/
+local instance : Module A A := Semiring.toModule
```

Comment:
I'd really like to know why these are necessary and to be able to drop them.

Reply:
Should be related to leanprover/lean#2074. Please mention that issue in the porting note.

Reply:
Ah indeed, I note that these instances are not necessary if I instead write:
```lean
set_option synthInstance.etaExperiment true
```

Reply:
So this is almost certainly caused by https://github.com/leanprover/lean4/issues/2074